# HTS Processing for the KatzLab Phylogenomic Pipeline

Tools to Process HTS data in conjunction with the KatzLab Phylogenomic Pipeline

### Update Note -- 06-Feb-2018

Currently working on an improved version of the work-flow (based on some requests)

Similarly, I am working on updating the appropriate links (some of which appear dead) to be more permanent (thanks for your patience).

### For germline-soma distinctions in ciliates

These scripts, for transcriptome processing, have been used in conjuction with bioinformatic steps dedicated to describing germline-soma genome architectures in ciliates [Germ-Soma Scripts - GitHub](https://github.com/maurerax/KatzLab/tree/SingleCellGermSoma)

Use the above link to learn more about the intent (and limitations) behind those steps.

### For more information:
Request viewing access to [Katz-Lab SciWiki](https://sciwiki.smith.edu/katzlab_phylo/index.php/Main_Page) from current members of the Katz Lab at Smith College. 

There you will find detailed information about each major script (step) as well as more detailed information on assembler choice and the phylogenomic pipeline as well. 


#### Note about PostAssembly.tar.gz 

This pre-zipped file contains the folder structure and scripts ready for use and broken down by data type.


#### Databases for PostAssembly.tar.gz

These can be downloaded from: ["Post-Assembly from GoogleDrive"](https://drive.google.com/drive/u/1/folders/0B7QiJDySGTwcSmtwNEdnUkpfUmc, "Post-Assembly from GoogleDrive")


## If using please cite:

INFO GOES HERE
